File: src/hyperLiquid_ts_deprecated/functions/executor.ts
import { SQSEvent, Context } from "aws-lambda";
import { services } from "../services/index";
import { WebhookPayload } from "../models/webhook";

const {
    parseWebhook,
    validateSignal,
    buildOrder,
    executeOrder,
    closeOrder
} = services;

/**
 * SQS-triggered Lambda to process trade signals.
 */
export const handler = async (event: SQSEvent, context: Context): Promise<void> => {
    try {
        console.log(`Processing ${event.Records.length} trade signals from SQS`);
        
        for (const record of event.Records) {
            const message = record.body;
            console.log("Raw SQS message:", message);
            const rawPayload = JSON.parse(message) as WebhookPayload;
            
            const payload = parseWebhook(rawPayload);
            
            // Context shim to maintain compatibility with existing Azure functions logs if relying on them inside services
            const mockContext = { 
                log: console.log, 
                error: console.error 
            };

            const validation = await validateSignal(payload, mockContext);
            console.log("Validation Result:", validation);

            if (!validation.isValid) {
                console.log(`Validation skipped: ${validation.reason}`);
                continue;
            }

            let orderResult: any;
            switch (payload.action.toUpperCase()) {
                case "ENTRY":
                    const tradeOrder = await buildOrder(payload, mockContext);
                    console.log("Built Order Request:", tradeOrder);
                    
                    // Uses core executeOrder logic preserving HL inputs
                    orderResult = await executeOrder(tradeOrder, mockContext);
                    console.log("Execute Order: ", JSON.stringify(orderResult));
                    break;

                case "EXIT":
                    orderResult = await closeOrder(payload, mockContext);
                    console.log("Close Order: ", JSON.stringify(orderResult));
                    break;

                default:
                    console.log(`Unknown action in queue: ${payload.action}`);
                    continue;
            }

            // Centralized tracking/logging mechanism
            await services.logTrade(payload, orderResult, mockContext);
        }
    } catch (error) {
        console.error(`Fatal error in background trade processing:`, error);
        throw error; // Will automatically let SQS handle dead letter DLQ/retries
    }
}

File: src/hyperliquid_python/functions/executor.py
import json
from services.parse_webhook import parse_webhook
from services.validate_signal import validate_signal
from services.build_order import build_order
from services.execute_order import execute_order
from services.close_order import close_order
from services.log_trade import log_trade

def handler(event, context):
    try:
        records = event.get('Records', [])
        print(f"Processing {len(records)} trade signals from SQS")
        
        for record in records:
            message = record.get('body')
            print("Raw SQS message:", message)
            
            raw_payload = json.loads(message)
            payload = parse_webhook(raw_payload)
            
            validation = validate_signal(payload)
            print(f"Validation Result: is_valid={validation.is_valid}, reason={validation.reason}, skipped={validation.skipped}")
            
            if not validation.is_valid:
                print(f"Validation skipped: {validation.reason}")
                continue
                
            action = payload.action.upper()
            order_result = None
            
            if action == "ENTRY":
                trade_order = build_order(payload)
                print(f"Build Order: {trade_order.symbol} {trade_order.type} {trade_order.quantity} @ {trade_order.price}")
                
                order_result = execute_order(trade_order)
                if order_result.success:
                    print(f"✅ Order Executed: {order_result.orderId}")
                else:
                    print(f"❌ Execution Failed: {order_result.error}")
                
            elif action == "EXIT":
                order_result = close_order(payload)
                if order_result.success:
                    print(f"✅ Exit Order Executed: {order_result.orderId}")
                else:
                    print(f"❌ Exit Failed: {order_result.error}")
                
            else:
                print(f"Unknown action in queue: {action}")
                continue
                
            if order_result:
                log_trade(payload, order_result)
                
    except Exception as error:
        print(f"Fatal error in background trade processing: {error}")
        raise error
